src/Chain.cpp

```diff
@@ -112,15 +112,60 @@ std::vector<int> Chain::ViterbiPath(int row, int col) const {
 /// column.
 IntVectorVector Chain::ViterbiPaths() const {
   IntVectorVector paths;
-  for (int i = 0; i < viterbi_idx_.rows(); i++) {
-    for (int j = 0; j < viterbi_idx_.cols(); j++) {
+  // Note that we call `viterbi_idx()` (not `viterbi_idx_`) because this ensures
+  // `viterbi_idx_` is computed.
+  for (int i = 0; i < viterbi_idx().rows(); i++) {
+    for (int j = 0; j < viterbi_idx().cols(); j++) {
       paths.push_back(ViterbiPath(i, j));
     }
   }
   return paths;
 };
 
 
+/// @brief If a Chain is clean, mark it as dirty and recursively mark the
+/// upstream Smooshishs as dirty.
+void Chain::MarkAsDirty() {
+  if (!is_dirty_) {
+    is_dirty_ = true;
+    prev_->MarkAsDirty();
+    curr_->MarkAsDirty();
+  }
+};
+
+
+/// @brief If a Chain is dirty, mark it as clean, resize its marginal/viterbi
+/// probability matrices to empty matrices, and recursively mark the upstream
+/// Smooshishs as clean.
+///
+/// Note that resizing a dirty Chain's marginal/viterbi probability matrices to
+/// empty matrices will force recomputations based on the upstream Smooshishs'
+/// probability matrices (see member functions `marginal()`, `viterbi()`, and
+/// `viterbi_idx()`).
+void Chain::MarkAsClean() {
+  if (is_dirty_) {
+    is_dirty_ = false;
+    marginal_.resize(0, 0);
+    viterbi_.resize(0, 0);
+    prev_->MarkAsClean();
+    curr_->MarkAsClean();
+  }
+};
+
+
+/// @brief If a Chain is dirty, recurse over the upstream Smooshishs looking for
+/// dirty Smooshables.
+/// @param[in,out] dirty_smooshables
+/// A vector of SmooshishPtr's storing dirty Smooshables.
+void Chain::AuxFindDirtySmooshables(
+    std::vector<SmooshishPtr>& dirty_smooshables) {
+  if (is_dirty_) {
+    prev_->AuxFindDirtySmooshables(dirty_smooshables);
+    curr_->AuxFindDirtySmooshables(dirty_smooshables);
+  }
+};
+
+
 // Functions
```

Comment:
See https://github.com/matsengrp/linearham/issues/39#issuecomment-277389890 for an example of Chain cleaning.